lib/task_manager.js

```diff
@@ -0,0 +1,66 @@
+class TaskManager {
+  constructor() {
+    this.taskData = {};
+    this.queue;
+    this.buffer = [];
+    this.lastPushedToBuffer = 0;
+  }
+
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', (data) => this.handleNewData(data));
+  }
+
+  clearQueue() {
+    this.queue.pause();
+    this.queue.clear();
+  }
+
+  #generateDefensiveCopy() {
+    const bufferCopy = [];
+    for (const data of this.buffer) bufferCopy.push(data);
+    return bufferCopy;
+  }
+
+  #shiftOriginalBuffer(shiftsCount) {
+    while (shiftsCount >= 0) {
+      this.buffer.shift();
```

Comment:
What's that shenanigan with shifting by one. Just erase the whole buffer by setting length to 0.

Reply:
If you do that, you risk losing data. 
Let's say we have a buffer of [1,2,3,4]. You copy this. In the meantime, between the copy and the buffer deletion, another element is added at the back, let's say 5. Then you'll erase the whole buffer
Though, I think this could be improved with Array.slice, now that you say it. Maybe it'd be a bit more optimal. But a shenanigan is indeed needed

Reply:
>  In the meantime, between the copy and the buffer deletion, another element is added at the back

What do you mean, NodeJS is single threaded, this is the whole proposition. Also if it wasn't (we were dealing with some other language) this approach to access the same data structure most probably wouldn't have been enough to give us thread safety.